client/repo.c

```diff
@@ -467,6 +467,155 @@ TDNFEventRepoMDDownloadEnd(
+    {
+        dwError = ERROR_TDNF_NO_REPOS;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    pRepos = pTdnf->pRepos;
+
+    while (pRepos)
+    {
+        if (!strcmp(pszRepo, pRepos->pszId))
+        {
+            break;
+        }
+        pRepos = pRepos->pNext;
+    }
+
+    dwError = TDNFAllocateStringPrintf(&pszBaseUrlFile,
+                                       "%s/%s/tmp/%s",
+                                       pTdnf->pConf->pszCacheDir,
+                                       pRepos->pszId,
+                                       TDNF_REPO_BASEURL_FILE_NAME);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+    dwError = TDNFCreateAndWriteToFile(pszBaseUrlFile, pszRepoMDURL);
+    BAIL_ON_TDNF_ERROR(dwError);
+
+cleanup:
+    TDNF_SAFE_FREE_MEMORY(pszBaseUrlFile);
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNFDownloadUsingMetalinkResources(
+    PTDNF pTdnf,
+    const char *pszRepo,
+    const char *pszFile,
+    const char *pszProgressData,
+    char **ppszRepoMDUrl,
+    TDNF_METALINK_FILE *ml_file
+    )
+{
+    uint32_t dwError = 0;
+    TDNF_METALINK_URLS *urls = NULL;
+    char *pszRepoMDUrl = NULL;
+    char buf[BUFSIZ] = {0};
+
+    if (!pTdnf ||
+        !pTdnf->pArgs ||
+        IsNullOrEmptyString(pszFile) ||
+        IsNullOrEmptyString(pszRepo))
+    {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+    urls = ml_file->urls;
+
+    while (urls)
+    {
+        dwError = TDNFStringEndsWith(urls->url, TDNF_REPO_METADATA_FILE_PATH);
+        if (dwError)
+        {
+            dwError = ERROR_TDNF_INVALID_REPO_FILE;
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+        dwError = TDNFDownloadFile(pTdnf, pszRepo, urls->url, pszFile,
+                                   pszProgressData, 0, NULL);
+        if (dwError)
+        {
+            urls = urls->next;
+            if (urls)
+            {
+                continue;
+            }
+            BAIL_ON_TDNF_ERROR(dwError);
+        }
+        strncpy(buf, urls->url, BUFSIZ-1);
+        buf[BUFSIZ-1] = '\0'; // force terminate
+        dwError = TDNFTrimSuffix(buf, TDNF_REPO_METADATA_FILE_PATH);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        dwError = TDNFStoreBaseURLFromMetalink(pTdnf, pszRepo, buf);
+        BAIL_ON_TDNF_ERROR(dwError);
+
+        dwError = TDNFAllocateStringPrintf(&pszRepoMDUrl,
+                                           "%s/%s",
+                                           buf,
+                                           TDNF_REPO_METADATA_FILE_PATH);
+        BAIL_ON_TDNF_ERROR(dwError);
+        *ppszRepoMDUrl = pszRepoMDUrl;
+        break;
+    }
+
+cleanup:
+    return dwError;
+error:
+    goto cleanup;
+}
+
+uint32_t
+TDNFFreeMetalinkUrlsList(
+    TDNF_METALINK_URLS *metalink_urls
+    )
+{
+    uint32_t dwError = 0;
+    TDNF_METALINK_URLS *metalink_urls_temp = NULL;
+    TDNF_METALINK_URLS *metalink_urls_temp_next = NULL;
+
+    if (!metalink_urls)
```

Comment:
metalink_urls list can be empty? which is a valid case. So returning 0 should be fine